Upgrade to material-ui v4

4ian opened the pull request:
This entirely remove material-ui v0 and replace everything with material-ui v4. Most components have been extracted to the `newIDE/app/src/UI` folder to have proper typing.

A few changes:
- Some components design slightly changed
- Buttons are smaller
- All lists should be displayed in the "dense" setting, so that they show more items
- Toolbar is a bit more dense too
- Toolbar have tooltips
- Snackbar (the large tips shown at the bottom) have been reworked so they are really dismissable: if you click on "Got it", they won't reappear again and again (unless you activate it again from preferences)

Still to do:

- [x] Autocomplete
  - [x] `focus()` is not working => missing focus when clicking for example on an object parameter in the events sheet.
- [x] Select field when an improper value is entered

Let me know if you're finding something broken or changed in an unexpected way. 
@Bouh I know that you started to look at this already when it's was still in progress - you can now take a full look ;)

Bouh commented:
I've tested a lot of things and considering the number of important changes nothing seems broken.
The rest of my post is in French, it was faster and easier for me.


- Les boutons avec un fond de couleurs ont des angles arrondi (Bouton page de démarrage, Edit points, edit hitboxes, Cancel, Apply, etc.)

- Le curseur avec la main sur les boutons, c'est déroutant parfois le curseur semble être à côté, en tout cas sur les contours du bouton mais pas sur le texte, c'était un peu le cas aussi sur la version actuelle mais là c'est beaucoup plus présent, on vois parfois la mains mais jamais entièrement sur les boutons.
Un exemple de bouton qui fonctionne très bien en version actuelle stable est le bouton Bold et Italic des objet textes

- Il n'est plus possible sur certain champs de valeurs numérique de scroller de haut en bas pour changer la valeur, impossible dans les propriétés de l'objet sur la scène (z-order, positions, angles, etc..), mais possible dans les animations sprite là on peut ajuster le temps de l'animation.
Impossible dans la size d'un objet texte, mais possible dans l'ajustement de la grille de la scène.

- Dans les propriété d'un objet texte, le color picker ne se ferme pas si on clique dans le champs pour saisir le texte de l'objet qui est en dessous du color picker

- Les signes d'attribution sont afficher bizarrement dans l'éditeur de condition : 
\u2264 (less or equal to)
\u2265 (greater or equal to)
\u2260 (not equal to)

- Dans l’éditeur de condition, choisi "Compare the two numbers"
Entre la première et la seconde expression. Ne touche pas au signe d'attribution qui est déjà sur égal.
Fait OK, dans la feuille le égal = ne sera pas visible.

- Les événements de bloc avancé, Repeat, Link, For each object, ne sont pas configurable.

- Les champs de sélection de fichier, comme choisir un fichier audio pour l'action "Play sound"
Il y a bien la possibilité de choisir un fichier, par contre il y a la scroll bar sur la droite qui ne sert à rien car il y a que un choix ou peut de choix. Elle devraient être visible uniquement après 5 choix par exemple ?

- Dans le même panel, pour le volume et le pitch, c'est un exemple, mais il est indiqué une valeur par défaut dans la description, cette valeur n'est pas défaut dans le champs, le champs est vide.

- Dans l'ensemble aucun bouton ne semble brisé, il y a juste du CSS à refaire.
L'interface est très dense par endroit, un peut d'espacement/séparation serait bienvenue.

- L'effet sur le bouton de fermeture passe par dessus.
En ouvrant le Project manager, au survol du bouton pour fermé le panel, l'effet passe par dessus, dans la version actuelle l'effet est restreint dans le cadre.

- Il n'y a plus d'indentation visuel des items dans la liste "Game settings", c'était une bonne chose sur la version actuelle, j'irais même jusqu'a étendre cette indentation pour tout, les scènes, les external event, layout, fonctions/behaviors. Histoire d'avoir une meilleur visibilités entre chaque catégories.

En haut la stable en bas cette branche material ui v4
![image](https://user-images.githubusercontent.com/1670670/64699106-62debb80-d4a4-11e9-8ae8-f60cc62db4c9.png)

- La polices d'écriture semblent être en gras, elle semblent ne pas être de même taille aussi.
Edit with piskel, preview rendent mieux dans la version stable.

![image](https://user-images.githubusercontent.com/1670670/64698205-b94afa80-d4a2-11e9-9899-f8cebc6ecc82.png)

![image](https://user-images.githubusercontent.com/1670670/64698371-0b8c1b80-d4a3-11e9-9893-305b38e8eb5a.png)


- Le panneau des icônes du projet est mal lisible par rapport à la version actuelle, tout est très dense et les messages d'erreur en rouge sont confus, sur quelle ligne ils sont ?

- Les listes de valeurs/sélection on deux versions différentes, une liste que de valeurs et une liste de valeur plus la sélection d'un fichier, il faudrait garder selon moi un style communs aux deux type de listes. Je sais qu'une est custom et l'autre en fonction de l'os, mais celle custom est plus stylé et plus complète.

- Je ne sais pas si c'est le focus ou autre choses, mais avec le nouvel éditeur d’événement si on ouvre un évent pour le modifié les listes sont toutes en haut, il n'y a pas de centrage sur la condition/action et/ou l'objet, par contre la sélection y est si on scroll pour retrouvé l'action/condition/objet.

- Pour la condition "PlayerHitBox is in collision with Enemy", il n'y a aucun centrage et aucune catégorie d'ouverte correspondant à cette condition. Par contre si on ouvre la bonne catégorie on vois la selection de la condition.
C'est du au faite que cette condition est dans "Other Conditions", c'est le même comportement pour les "Other actions"

- Dans la toolbar avec les scènes et feuille d’événements, la dimension des onglets n'a plus une taille minimum, se qui est un peut dommage pour les fermer rapidement, je suggère le clique molette sur l'onglet pour le fermé, comme le fait Google Chrome.

- La fonction de recherche dans les extensions/fonctions fait n'importe quoi sur l'interface (Rapporté dans une précédente issue)

4ian commented:
> Le curseur avec la main sur les boutons [cursor showing hand/default alternatively on buttons]

Fixed this! Was very disturbing indeed, I completely missed this cursor related "detail" :)

> Ne touche pas au signe d'attribution qui est déjà sur égal. [select fields already showing a value while none is chosen]

Fixed

>  dimension des onglets n'a plus une taille minimum, [tab size]

I've added a right click menu with Close All and Close Others items, that would maybe be helpful? I can reconsider having a minimum size.

Bouh commented:
> > Ne touche pas au signe d'attribution qui est déjà sur égal. [select fields already showing a value while none is chosen]
> 
> Fixed

- But it is still possible to see an empty condition.
On the second condition there is a nice red line that indicates an error, it would be nice to have this on all fields where a value is missing, or on a field that contains text instead of a number, and vice versa.
As soon as a type of data is put in the wrong place, you see a red line.
It would educate people with types text, number, variables, scenes, key, mouse button, and it makes mistakes much more visible.

Maybe for future update, i know it's not related to this migration to v4 :)
I will add card for this on Trello
![image](https://user-images.githubusercontent.com/1670670/64901120-2ffc1980-d696-11e9-8dc8-78acefc8dda2.png)

> I can reconsider having a minimum size.

Not need anymore, we have a nice improvement for tabs now !

4ian commented:
I think I've fixed everything you and I could find, and from using the whole app I don't spot any more issues. I've re-implemented the scroll in the list to the selected action/condition when editing an action/condition. Should be good to go!

I'll merge now. If there are minor issues, I can fix them after. I don't think there will be a big regression, but if that the case let me know :)

Bouh commented:
Possible to add this ?
Rounded button are ugly, and texts are little bolder than before because material ui 4 configure the fontWeight to 500 instead 400.
These little details made the interface a little strange.
```
    MuiButton: {
      root: {
        borderRadius: 0,
        fontWeight: 400,
      },
    },
```

I note nothing blocking for use in production.
That sounds good to me too:)

4ian commented:
I got used to these but might be true that on Windows where things are more square now, buttons can be strange. I applied this on master :)

Bouh commented:
- Text in expressions list are black.
- In Hitbox and Edit point, text "Frame" is over "Choose an option", instead on top. Only when we have no frame. And we can use theses button hitbox and point we have frames in animations.
![image](https://user-images.githubusercontent.com/1670670/65325946-8692b780-dbb0-11e9-9bb0-a08f4618061f.png)

4ian commented:
Both should now be fixed, thanks for reporting this :)
